src/parse.rs

```diff
@@ -0,0 +1,319 @@
+use crate::config::{ParserSettings, LINK_REGEX};
+use crate::document::{
+    code::{CodeBlock, Line, Source},
+    text::TextBlock,
+    transclusion::Transclusion,
+    Document, Node,
+};
+use crate::util::{Fallible, TryCollectExt};
+use std::error::Error;
+use std::fmt::Write;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+#[allow(clippy::nonminimal_bool)]
+pub fn parse(input: &str, path: &Path, settings: &ParserSettings) -> Fallible<Document> {
+    #[derive(Default)]
+    struct State {
+        node: Option<Node>,
+    }
+
+    enum Parse {
+        Incomplete,
+        Complete(Node),
+        Error(Box<dyn Error>),
+    }
+
+    impl Parse {
+        fn error(err: Box<dyn Error>, line: usize) -> Self {
+            Self::Error(format!("{} (line {})", err, line).into())
+        }
+    }
+
+    let mut state = State::default();
+    let mut document = input
+        .lines()
+        .enumerate()
+        .scan(&mut state, |state, (line_number, line)| {
```

Comment:
While this looks nicely functional, I am not sure if implementing the parser via `scan` and `try_collect` is really helpful. I would suggest at least giving a plain for over the numbered lines that can manually aggregate the nodes and errors a try. (Especially as such a loop could just access `nodes.last_mut()` instead of `take()`ing and set `state.node`.)

On a related note, I suspect it would be helpful to factor out parts of this parser into separate functions as well afterwards.

Reply:
I will give it a try. I also think the use of `scan` makes it unnecessary hard to grasp.

Reply:
See #97.